org.eclipse.m2e.tests.common/src/org/eclipse/m2e/tests/common/AbstractMavenProjectTestCase.java

```diff
@@ -90,24 +104,100 @@
 import org.eclipse.m2e.core.project.MavenUpdateRequest;
 import org.eclipse.m2e.core.project.ProjectImportConfiguration;
 import org.eclipse.m2e.core.project.ResolverConfiguration;
+import org.eclipse.m2e.core.ui.internal.M2EUIPluginActivator;
 import org.eclipse.m2e.jdt.MavenJdtPlugin;
 import org.eclipse.m2e.jdt.internal.BuildPathManager;
 
 
 @SuppressWarnings("restriction")
 public abstract class AbstractMavenProjectTestCase {
 
+  private static final boolean STOP_ON_FAILED_TEST = Boolean.getBoolean("m2e.stopOnFailedTest");
+
   public static final int DELETE_RETRY_COUNT = 10;
 
   public static final long DELETE_RETRY_DELAY = 6000L;
 
-  protected static final IProgressMonitor monitor = new NullProgressMonitor();
+  protected static final IProgressMonitor monitor = new IProgressMonitor() {
+
+    AtomicBoolean canceled = new AtomicBoolean();
+
+    public void worked(int work) {
+      driveEvents();
+    }
+
+    public void subTask(String name) {
+      driveEvents();
+
+    }
+
+    public void setTaskName(String name) {
+      driveEvents();
+
+    }
+
+    public void setCanceled(boolean value) {
+      canceled.set(value);
+    }
+
+    public boolean isCanceled() {
+      return canceled.get();
+    }
+
+    public void internalWorked(double work) {
+      driveEvents();
+
+    }
+
+    public void done() {
+      driveEvents();
+
+    }
+
+    public void beginTask(String name, int totalWork) {
+      driveEvents();
+    }
+  };
 
   protected IWorkspace workspace;
 
   @Rule
   public TestName name = new TestName();
 
+  @Rule(order = Integer.MIN_VALUE)
+  public TestWatcher watchman = new TestWatcher() {
+    @Override
+    protected void failed(Throwable e, Description description) {
+      if(STOP_ON_FAILED_TEST) {
```

Comment:
Is this opening a shell to abort the test?
If yes, should this really be submitted to the master? If the test case tends to dead-lock/run forever I suggest to assign a timeout to it.

Reply:
This open a small Shell with a button to wait and keep the test-eclipse open so you are able to inspect the state (e.g. project, console outputs etc.) as the tests run in the main thread, using a breakpoint don't work because the test-eclipse is frozen as long as the main thread is suspended.

This has become quite handy for me to debug the tests, as it is sometimes quite hard to guess just from the object state whats going on.

Reply:
That's indeed a problem I just encountered yesterday.
What I have done in such case, is that I have restarted the test after I have changed the configuration to not run in the UI-thread. That may work for some tests but probably not for all.

So yes that can be handy. 👍🏽